sort.h

```diff
@@ -847,60 +854,324 @@ static void TIM_SORT_RESIZE(TEMP_STORAGE_T *store, const size_t new_size) {
   }
 }
 
-static void TIM_SORT_MERGE(SORT_TYPE *dst, const TIM_SORT_RUN_T *stack, const int stack_curr,
-                           TEMP_STORAGE_T *store) {
-  const size_t A = stack[stack_curr - 2].length;
-  const size_t B = stack[stack_curr - 1].length;
-  const size_t curr = stack[stack_curr - 2].start;
-  SORT_TYPE *storage;
-  size_t i, j, k;
-  TIM_SORT_RESIZE(store, MIN(A, B));
-  storage = store->storage;
 
-  /* left merge */
-  if (A < B) {
-    memcpy(storage, &dst[curr], A * sizeof(SORT_TYPE));
-    i = 0;
-    j = curr + A;
-
-    for (k = curr; k < curr + A + B; k++) {
-      if ((i < A) && (j < curr + A + B)) {
-        if (SORT_CMP(storage[i], dst[j]) <= 0) {
-          dst[k] = storage[i++];
-        } else {
-          dst[k] = dst[j++];
+static size_t TIM_SORT_GALLOP(SORT_TYPE *dst, const size_t size, const SORT_TYPE key, size_t anchor,
+                              int right) {
+  int last_ofs = 0;
+  int ofs, max_ofs, cmp;
+  size_t  l, c, r;
+  cmp = SORT_CMP(key, dst[anchor]);
+
+  if (cmp < 0 || (!right && cmp == 0)) {
+    /* short cut */
+    if (anchor == 0) {
+      return 0;
+    }
+
+    ofs = -1;
+    max_ofs = -anchor; /* ensure anchor+max_ofs is valid idx */
+  } else {
+    if (anchor == size - 1) {
+      return size;
+    }
+
+    ofs = 1;
+    max_ofs = size - anchor - 1;
+  }
+
+  for (;;) {
+    /* deal with overflow */
+    if (max_ofs / ofs <= 1) {
+      ofs = max_ofs;
+
+      if (ofs < 0) {
+        cmp = SORT_CMP(key, dst[0]);
+
+        if ((right && cmp < 0) || (!right && cmp <= 0)) {
+          return 0;
         }
-      } else if (i < A) {
-        dst[k] = storage[i++];
       } else {
+        cmp = SORT_CMP(dst[size - 1], key);
+
+        if ((right && cmp <= 0) || (!right && cmp < 0)) {
+          return size;
+        }
+      }
+
+      break;
+    }
+
+    c = anchor + ofs;
+    /* right, 0<ofs:  dst[anchor+last_ofs] <=  key  <  dst[anchor+ofs]      */
+    /* left,  0<ofs:  dst[anchor+last_ofs] <   key  <= dst[anchor+ofs]      */
+    /* right, ofs<0:  dst[anchor+ofs]      <=  key  <  dst[anchor+last_ofs] */
+    /* left,  ofs<0:  dst[anchor+ofs]      <   key  <= dst[anchor+last_ofs] */
+    cmp = SORT_CMP(key, dst[c]);
+
+    if (0 < ofs) {
+      if ((right &&  cmp < 0) || (!right && cmp <= 0)) {
+        break;
+      }
+    } else {
+      if ((right && 0 <= cmp) || (!right && 0 < cmp)) {
         break;
       }
     }
+
+    last_ofs = ofs;
+    ofs = ofs << 1 + 1;
```

Comment:
This produces the warning:

    warning: operator '<<' has lower precedence than '+'; '+'
    will be evaluated first [-Wshift-op-parentheses]

Did you mean `(ofs << 1) + 1`?